lib/ansible/module_common.py

```diff
@@ -119,7 +119,22 @@ def _find_snippet_imports(self, module_data, module_path):
                 snippet_name = tokens[2].split()[0]
                 snippet_names.append(snippet_name)
                 output.write(self.slurp(os.path.join(self.snippet_path, snippet_name + ".py")))
-
+            elif line.startswith('from ansible.module_custom.'):
+                module_custom_path = C.DEFAULT_ANSIBLE_CUSTOM_MODULES
+                if module_custom_path is None:
+                    import_error = True
+                    raise errors.AnsibleError("ANSIBLE_CUSTOM_MODULES is None: %s" % C.DEFAULT_ANSIBLE_CUSTOM_MODULES)
+                tokens=line.split(".")
+                import_error = False
+                if len(tokens) != 3:
+                    import_error = True
+                if " import *" not in line:
+                    import_error = True
+                if import_error:
+                    raise errors.AnsibleError("error importing module in %s, expecting format like 'from ansible.module_utils.basic import *'" % module_path)
+                snippet_name = tokens[2].split()[0]
+                snippet_names.append(snippet_name)
+                output.write(self.slurp(os.path.join(module_custom_path, snippet_name + ".py")))
```

Comment:
>the name custom_modules will confuse people, these are really module copied requirements or dependencies

Please suggest the name of Your choice.
What about custom_lib_module?

>I would also look at plugin loader to allow for local paths, the way other plugins do, this way a module that comes with a play or role can have their extra dependencies also provided vs needing to install them in a specific ansible location.

I think there is a question: how deep should we include dependences into single merged_module_file. I would suggest, that the 'standard python packages' should be installed (or manually delivered if needed) into remote host environment. 

>this assumes that the direct import will be the only thing needed, what about dependencies of that library?

That's true, for now it does.
I did just extend the current ansible functionality (where modules are almost self-contained).

For now, I do assume, that the required custom modules (and their dependencies) would be provided in the ANSIBLE_CUSTOM_MODULES path. 

For custom_lib_module depending on another custom_lib_module:
Let's make an example: a lib file (in example: tools.py) depends on the other lib file (conf.py), then if You would import both into Your module (websrv), the ModuleReplacer would merge content of all files (websrv, tools.py, conf.py) into one, and the dependence: tools.py -> conf.py  would be resolved by this phenomena.

I do assume, that the other dependencies would be just a python packages - that should be installed on the remote hosts (this can be done by other play).

>should we allow for a path or archive to copy?

I'm not sure if I do understand this part of the question. 

-------------------
IDEA
We could also think about using the eggs.
Provided we store them in customized location on local host, we have a few ways of importing and egg into our merged_module_file.
We can: 
- append eggs into merged_module_file (and loading them on_the_fly) 
- sending them to the remote host before running the module, and import their content